Alert management

maxbritto opened the pull request:
- Adds a function to allow automatic dismissal of alerts that prevent
  actions or snapshots
- Adds the possibility to wait for any alert to be dismissed before a
  snapshot is taken (disabled by default)

maxbritto committed:
Merge remote-tracking branch 'fastlane/master' into develop

* fastlane/master: (254 commits)
  Fixed quotation marks in Quick Start description.
  Code style fix
  Use `to_store.count` instead of `count`
  Use type: Array for options which expect Array values
  Remove unused `activities` variable
  Reduce iterations in activity finder
  Enabled UI layer
  Updated error message
  Improved debug information on error
  Fixed code style
  Version bump
  Add resetting of TVs and watches
  Updated snapshot warning
  Updated SnapshotHelper to work with legacy code
  Added 'match'
  Version bump
  Made app identifier optional
  Renamed PEM to pem
  Updated docs
  Version bump
  ...

maxbritto committed:
Alert amangement

- Adds a function to allow automatic dismissal of alerts that prevent
actions or snapshots
- Adds the possibility to wait for any alert to be dismissed before a
snapshot is taken (disabled by default)

maxbritto commented:
This PR is following this one : https://github.com/fastlane/snapshot/pull/289

ohwutup commented on lib/assets/SnapshotAlertManager.swift:
```diff
@@ -0,0 +1,60 @@
+func waitForAlertsToBeDismissed() {
+    SnapshotAlertManager.waitForAlertsToBeDismissed()
+}
+
```
I'm just a bit curious why these are global functions. Should/could they be class functions of SnapshotAlertManager?

maxbritto replied:
I followed the same format as the current SnapshotHelper.swift file. This allows the methods to be called without prefixing the call with SnapshotAlertManager.
But you are right about the global variable though, I changed it to a type property, I just have to figure out how to update the PR (I've always used subversion and I'm still discovering git)

ohwutup replied:
No worries! You should be able to just push up to the same branch!

maxbritto replied:
Yep, I forgot to click "sync" on the github app after my commit. It worked like a charm after I did :)

ohwutup commented:
Thanks @maxbritto ! This is awesome! 

Just one question above and then we can :shipit: !

maxbritto committed:
Alert management

- changed alertInterruptionMonitor from a global variable to a type
property

maxbritto commented:
Ok I've updated the code to address your question. I kept it the default behavior to not wait for alerts before a snapshot but it's just a boolean to enable the wait. This is the same behavior as with the 'waitForLoadingIndicator' parameter. But since @KrauseFx told to enable it by default if it works I don't know what to do

ohwutup commented:
You can enable it by default if you'd like, especially given @KrauseFx's endorsement.

maxbritto committed:
Alert management

- making the snapshot function waiting for alerts to be dismissed as
the default behavior.

maxbritto committed:
Alert Management

- displaying the automaticallyDismissAlerts recommandation only if we
actually have been waiting for an alert for more than 2 seconds
- generate a fake action to hide the alert only if the
automaticallyDismissAlerts feature has been activated

maxbritto commented:
Ok I did, but I left the automatic dismissal of alerts as an opt-in because I suppose some developer may want to control what to do with their alerts instead of dismissing them. I added some documentation to explain how to use it if needed (as logging in the setup.rb and as and when waiting for an alert for too long)

ohwutup commented on lib/assets/SnapshotAlertManager.swift:
```diff
@@ -0,0 +1,60 @@
+    class func automaticallyDismissAlerts(testCase:XCTestCase) -> NSObjectProtocol
+    {
+        SnapshotAlertManager.stopDismissingAlerts(testCase)
+        alertInterruptionMonitor = testCase.addUIInterruptionMonitorWithDescription("Snapshot alert auto hide") { (alert: XCUIElement) -> Bool in
```
I'm not familiar with how this `addUIInterruptionMonitorWithDescription` works, but how does this determine which button on an alert it should tap?

maxbritto replied:
It is one of those undocumented functions... :( 
I've seen many code examples which were manually touching one button by typing its title and then returning true to indicate the event had been consumed. But if you do nothing and return false the alert one button is clicked automatically. I've tested it on one of my own alerts with only one button and it was clicked and for system alerts (push notification user authorization) it is the 'OK' button that is clicked for me.

ohwutup replied:
well the lack of docs is awful, but the functionality is sweet. is that how `XCUIApplication().tap()` works below as well?

maxbritto replied:
I've discovered that the UIInterruptionMonitor event is not triggered when the alert is showed but when the alert is blocking an action. So if an alert is displayed and no action is done, the alert will stay here forever. Hence my `XCUIApplication().tap()` after 2 seconds of waiting to trigger the method callback that will return false and close the alert.

ohwutup replied:
:rocket:

maxbritto replied:
Cool :+1:

ohwutup commented:
okay, so final thing. would you mind squashing the commits? i know you're new to git, so here are some rough instructions:

`$ git rebase -i HEAD~5`
you'll now be in an interactive rebase, you'll be asked how to handle each commit, and can choose `fixup` for all commits except for the first commit on this branch. on the first commit in this branch, choose `reword` and you'll be prompted to enter a new commit message for the single commit that will represent the branch.

maxbritto commented:
<img width="594" alt="capture d ecran 2016-01-07 a 23 14 24" src="https://cloud.githubusercontent.com/assets/1722050/12184257/7462c410-b594-11e5-9368-0d168dc64a46.png">
Ok, just to be sure, I have 258 commands that are by default on `pick` I have to search/replace it by `fixup` and replace only the one for the `e3c148e` by `reword` ?

ohwutup commented:
wah, okay. so there looks like there is a bit of an issue because there was a merge commit first, was your first commit `e3d148e` ?

maxbritto commented:
Yes it is the first one of four.

ohwutup commented:
ah, okay. so we will have to do a bit more surgery. 

i think you should try to do the following.

first- check out a new branch from where you are `git checkout -b backup`, to back it up
go back to this branch and then do a `git reset e3d148e`
that will bring you back to the beginning of your changes, then do `git stash` to hold on to the changes for a few minutes
then do `git checkout master` and then `git pull https://github.com/fastlane/snapshot.git master` to get the most recent changes on master
delete this branch with `git branch -D <this branch name>`
the do `git checkout -b <this same branch name>` followed by `git stash pop` to get your changes back
finally, commit those changes and push again

ohwutup commented:
ah, this is a bit tougher than i though. i am pulling down the changes and seeing what i can do

maxbritto commented:
Or I can save my 3 files, delete my branch an cancel this PR then do a new one with one commit ? We'll loose the benefit of this discussion though

ohwutup commented:
thats not a bad idea, but if you call your new branch, the same thing as this one, i _think_ it should update it.

maxbritto commented:
I tried to do something that ended up in https://github.com/fastlane/snapshot/pull/378 but I have the feeling that it wasn't the right way since I see the commit from november in it :(

KrauseFx commented:
Should this be closed in favor of the other PR?

maxbritto commented:
@KrauseFx : probably, I'm doing it now
